versioning/create-release.py

```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/env python3
+from sh.contrib import git
+from sh import cat
+import os, sys, argparse, sh
+
+version_file="VERSION"
+
+# TODO implement
+def update_self(branch):
+    previous_dir = os.getcwd()
+    path, file = os.path.split(os.path.realpath(__file__))
+    print("Update {} in {}".format(file, path))
+    os.chdir(path)
+    if branch == "":
+        branch = 'master'
+    check_branch(branch)
+    repo_is_clean()
+    gitcmd('fetch', 'git fetch failed')
+    gitcmd('pull', 'git pull failed')
+    os.chdir(previous_dir)
+    print("Release script updated.")
+    return
+
+# print error message and exit
+def error_exit(errMsg):
+  print("Error: {}".format(errMsg))
+  sys.exit(1)
+
+# write the version string to the version file
+def write_version(ver_str):
+    try:
+        f = open(version_file, "w")
+        f.write(ver_str)
+        f.close()
+    except:
+        error_exit('unable to write version to file')
+
+# wrapper function for calling git commands
+def gitcmd(command, error_string):
+    try:
+        res = git(command.split(' ')).strip('\n')
+    except:
+        error_exit(error_string)
+    return res
+
+# check for untracked files or uncommitted changes
+def repo_is_clean():
+    res = gitcmd('status --porcelain', 'git status failed')
+    if res != '':
+        error_exit('uncommitted changes or untracked files present')
+
+# check if current branch matches the requested branch
+def check_branch(branch):
+    res = gitcmd('rev-parse --abbrev-ref HEAD', 'git rev-parse failed')
+    if (res != branch):
+        error_exit('current branch {} != {}'.format(res, branch))
+
+def git_checkout_branch(branch):
+    gitcmd('checkout {}'.format(branch), 'unable to checkout branch {}'.format(branch))
+
+def git_checkout_create_branch(branch):
+    res = gitcmd('checkout -b {}'.format(branch), 'unable to checkout/create branch {}'.format(branch))
+    print(res)
+
+# Return tags as a string, replace newlines with space
+def git_get_tags():
+    return gitcmd('tag', 'failed to get tags').replace('\n', ' ')
+
+
+# used for both tags and VERSION file, strips leading 'v'
+# 'v1.2.3' -> [1, 2, 3] and '1.2.3' -> [1, 2, 3]
+def version_from_string(ver_str):
+    if len(ver_str) < 5:
```

Comment:
Why are we checking the length of the string here? What if there is a (manually created) version string that looks like this: `v1.2`?

Reply:
just wanted to be sure that the string could at least contain 0.0.0
might switch to regex

Reply:
switched to regex